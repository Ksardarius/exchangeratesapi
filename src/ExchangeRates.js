import React, { useState } from 'react';
import { gql } from 'apollo-boost';
import { useQuery } from '@apollo/react-hooks';
import Loader from 'react-loader-spinner'
import { get, keys } from 'lodash/fp'

export const EXCHANGE_RATES = gql`
  query exchangeRates($skip: Boolean!) {
    exchangeRates @client @skip(if: $skip) {
      base
      date
      rates
    }
  }
`;

const ExchangeRates = ({onError}) => {
  const [loadData, setLoadData] = useState(false)

  const { loading, error, data } = useQuery(EXCHANGE_RATES, {
    variables: {
      skip: !loadData,
    }
  });

  if (loading) {
    return <div className="loader"><Loader
      type="Puff"
      color="#00BFFF"
      height={100}
      width={100}
    /></div>
  }

  if (error) {
    onError && onError(error.message, 'Error');
  }

  return <div className="container mt-5">
    <div className="row">
      <button type="button" className="btn btn-primary btn-lg btn-block" onClick={_ => {setLoadData(true)}}>
        Get rates
      </button>
    </div>
    {data && data.exchangeRates && (<React.Fragment>
      <div className="row mt-3" data-testid="exchange-main-row">
        <div className="col-sm-6">
          <span>Base: <strong>{get('exchangeRates.base', data)}</strong></span>
        </div>
        <div className="col-sm-6">
          Date: <strong>{get('exchangeRates.date', data)}</strong>
        </div>
      </div>
      <div className="row mt-3">
        {get('exchangeRates.rates', data) && (

          keys(data.exchangeRates.rates).map((key, i) => (
            <div key={i} className="col-sm-3">
              {key}:<strong className="pl-1">{data.exchangeRates.rates[key]}</strong>
            </div>
          ))

        )}
      </div>
    </React.Fragment>)}
  </div>
}

export default ExchangeRates