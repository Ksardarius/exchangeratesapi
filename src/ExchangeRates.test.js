import React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import ExchangeRates, { EXCHANGE_RATES } from './ExchangeRates'
import { render, fireEvent, waitForElement } from '@testing-library/react'
import { NotificationContainer } from 'react-notifications';
import wait from 'waait';

const mocks = [
  {
    request: {
      query: EXCHANGE_RATES,
      variables: {
        skip: false,
      },
    },
    result: {
      data: {
        "exchangeRates": {
          "base": "EUR",
          "date": "2019-11-22",
          "rates": {
            "CAD": 1.4679,
            "HKD": 8.6525
          },
        }
      },
    },
  }
];

it('initially there should be a button', async () => {
  const { getByText, asFragment } = render(
    <MockedProvider mocks={[]} addTypename={false}>
      <ExchangeRates />
    </MockedProvider>
  )

  expect(getByText('Get rates')).toBeDefined()
});

it('"Loading" UI component should be shown while data is fetched', async () => {
  const { getByText, asFragment } = render(
    <MockedProvider mocks={[]} addTypename={false}>
      <ExchangeRates />
    </MockedProvider>
  )

  fireEvent.click(getByText('Get rates'))

  expect(asFragment()).toMatchSnapshot()
});

it('data is fetched when the button is clicked', async () => {
  const { getByText, getByTestId, asFragment } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <ExchangeRates />
    </MockedProvider>
  )

  fireEvent.click(getByText('Get rates'))
  await wait(0);

  await waitForElement(() => getByTestId('exchange-main-row'))

  expect(asFragment()).toMatchSnapshot()
});

it('error modal should be shown when data fetch fails', async () => {
  const errorMocks = [
    {
      request: {
        query: EXCHANGE_RATES,
        variables: {
          skip: false,
        },
      },
      error: new Error('Service error'),
    }
  ];

  const onError = jest.fn()

  const { getByText } = render(
    <MockedProvider mocks={errorMocks} addTypename={false}>
      <React.Fragment>
        <NotificationContainer />
        <ExchangeRates onError={onError} />
      </React.Fragment>
    </MockedProvider>
  )

  fireEvent.click(getByText('Get rates'))

  await wait(0);

  expect(onError).toHaveBeenCalled()
  expect(onError.mock.calls[0][0]).toBe('Network error: Service error');
  expect(onError.mock.calls[0][1]).toBe('Error');
});
