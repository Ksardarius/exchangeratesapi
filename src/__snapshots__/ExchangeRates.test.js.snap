// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"Loading" UI component should be shown while data is fetched 1`] = `
<DocumentFragment>
  <div
    class="loader"
  >
    <div
      aria-busy="true"
      class=""
    >
      <svg
        aria-label="audio-loading"
        height="100"
        stroke="#00BFFF"
        viewBox="0 0 44 44"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
          stroke-width="2"
        >
          <circle
            cx="22"
            cy="22"
            r="1"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="strokeOpacity"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
          <circle
            cx="22"
            cy="22"
            r="1"
          >
            <animate
              attributeName="r"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="strokeOpacity"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`data is fetched when the button is clicked 1`] = `
<DocumentFragment>
  <div
    class="container mt-5"
  >
    <div
      class="row"
    >
      <button
        class="btn btn-primary btn-lg btn-block"
        type="button"
      >
        Get rates
      </button>
    </div>
    <div
      class="row mt-3"
      data-testid="exchange-main-row"
    >
      <div
        class="col-sm-6"
      >
        <span>
          Base: 
          <strong>
            EUR
          </strong>
        </span>
      </div>
      <div
        class="col-sm-6"
      >
        Date: 
        <strong>
          2019-11-22
        </strong>
      </div>
    </div>
    <div
      class="row mt-3"
    >
      <div
        class="col-sm-3"
      >
        CAD:
        <strong
          class="pl-1"
        >
          1.4679
        </strong>
      </div>
      <div
        class="col-sm-3"
      >
        HKD:
        <strong
          class="pl-1"
        >
          8.6525
        </strong>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
